### ISSUE TITLE
Use notify url to check payment response

### ISSUE THREAD

**Issue by sanderha** [thephpleague/omnipay#344]
Hello

I am trying to write a driver for the Quickpay gateway (https://quickpay.net//). I'd use an existing one, but I dont think there is one.
I've set it up correctly so far, so that I can make an authorize at the gateway, so reserve some money on the customers card.

I am missing how to get information back about the payment. Quickpay does not include these things in their response, but to a callbackURL instead. 
I can understand there is something called a notifyUrl in Omnipay but I can't figure out how to use this to check the response on the callback url.

So my question is basicly; How do I check the response of the off-site payment, when it is not send to getReturnUrl() but to getNotifyUrl() instead? 

The not-finished code for my driver can be seen here: https://github.com/NobrainerWeb/omnipay-quickpay

All help and guidelines are very much appreciated!

**Comment by judgej** [comment]
tl;dr: bug in memcached cacheing plugins meant that transients were not being 
cached.

I would still question why the WC site was asking for 2500 database queries for 
the category pages in the first place!


On 25/02/2016 11:00, sanderha wrote:
>
> Hello
>
> I am trying to write a driver for the Quickpay gateway 
> (https://quickpay.net//). I'd use an existing one, but I dont think there is one.
> I've set it up correctly so far, so that I can make an authorize at the 
> gateway, so reserve some money on the customers card.
>
> I am missing how to get information back about the payment. Quickpay does not 
> include these things in their response, but to a callbackURL instead.
> I can understand there is something called a notifyUrl in Omnipay but I can't 
> figure out how to use this to check the response on the callback url.
>
> So my question is basicly; How do I check the response of the off-site 
> payment, when it is not send to getReturnUrl() but to getNotifyUrl() instead?
>
> The not-finished code for my driver can be seen here: 
> https://github.com/NobrainerWeb/omnipay-quickpay
>
> All help and guidelines are very much appreciated!
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/thephpleague/omnipay/issues/344>.
>


-- 
Jason Judge, Director
*Academe Computing Ltd*

Office: 0191 251 0135 (with voice mail)
Mobile: 07771 656 448
14 Collingwood Terrace, Whitley Bay, NE26 2NP, UK
Registered in England No. 3205068; VAT Reg. No. 647 6432 17

jason.judge@academe.co.uk <mailto:jason.judge@academe.co.uk>
academe.co.uk <http://academe.co.uk/>

**Comment by sanderha** [comment]
@judgej Did you mean to comment on another issue?

**Comment by judgej** [comment]
Haha! Sorry. No idea how that happened. I replied by email to an email from a colleague about something else altogether. And somehow it ended up here!? I must have slipped when hitting reply. I'll get me coat...

**Comment by judgej** [comment]
Hopefully I can add something to help here.

The `notifyUrl` is the `callback` URL that the remote gateway uses to notify your application of the transaction result. You will have a route at that URL to handle that response. Part of handling it will be to store the result of the transaction in the database, but only after checking the notification is from the trusted source (a hash is often used with a shared secret that itself is *never* sent between the application and gateway).

A very simple example of a notify handler for Sage Pay can be found here:

https://github.com/academe/OmniPay-SagePay-Demo/blob/master/sagepay-confirm.php

It uses the `completePurchase()` method to capture and validate the details sent from the gateway, and the `ServerCompleteAuthorizeResponse` message to return the result to let Sage Pay know whether we accept what was provided or not.

So, because this notification from the gateway is not in the user's session, it has no knowledge which transaction is being processed. So you need to store that in the database against the `transactionId` as a key. You give the `transactionId` to the gateway so that when it does the callback, it also passes back the `transactionId` so that you can use that to get the transaction out of the database so that it can be updated with the details of the authorisation.

When the user is returned to your site, the `transactionId` will be in their session still, and you can use that to look up the transaction in the database again to get the result then take appropriate action.

Some gateways do not provide enough details even in the Notification (e.g. Helcim, where you don't actually know how much was authorised). But it should give you a `transactionReference` at least so that you can then do another call to the gateway to get the full details, address, delivery notes etc. that will have been posted by the end user to the gateway site.

Does that provide enough context to how gateways that use callbacks work?

**Comment by sanderha** [comment]
Hehe no problem about your first comment being posted on the wrong issue, shit happens!

--

Thanks for your response, it explained quite a lot, and will most likely help me in getting further with my driver.

Would you mind taking a quick look at the process I am running for my driver, to see if it makes any sense? I am still wrapping my mind around the entire Omnipay process, so it could be great with a little feedback.

--

**Currently it is set up like this:**

First, I call ``` authorize() ``` (which is basicly the same as ``` purchase() ``` on my gateway. This then checks the data to send.

It then uses the ``` send() ``` method, to put this data into a ``` PurchaseResponse ``` , which then creates the RedirectURL, to send the user to an off-site payment form.

When the user then comes back to my website from the off-site payment ``` CompletePurchaseRequest ``` is used. Here I use ``` $this->httpRequest->query->all() ``` to send as data into  ``` CompletePurchaseResponse ``` 

Currently the ``` isSuccessful ``` method on  ``` CompletePurchaseReponse ``` just returns true, because I dont get a response I can validate the response with. This is basicly where the issue with callback comes in.

--

The way this is set up, does it seem like the right way to do it?

Thanks a lot.

**Comment by delatbabel** [comment]
Hi,

What you have described above is a good example of a handler for returnUrl.  As a summary, your control flow works like this:

* call `authorize()` (or maybe `purchase()`).
* get a response. For a redirect gateway this will be a RedirectResponse
* check for success.  If it's a success and a RedirectResponse then redirect the customer to the off-site payment form.
* When the user then comes back to your website call `completePurchase()`.

All of this is as you have described above. Have a look in the class documentation for omnipay-paypal RestPurchaseRequest for another example along similar lines.

Now your problem is that *subsequent to* [note 1] the redirect, you will get a notify sent to your notifyUrl. Not all gateways do this but clearly Quickpay does as does SagePay.  Some others that do include MultiCards and PaymentWall.

This message sent to your notifyUrl is a *POST* message containing a chunk of data. The content of that chunk of data varies from gateway to gateway but usually includes at least a status message (to update the status of the purchase) and some other data.

Your application, not your QuickPay driver, needs to be able to handle this POST request. Clearly Omnipay can't handle it because Omnipay is not an application, does not have a router and can't handle requests, it's just a data transmission layer.  It would be helpful if the documentation in your QuickPay driver contained the structure of this POST request so that anyone implementing QuickPay via Omnipay has that documented.

I will say in reference to @judgej 's example for SagePay, it's very unusual for a gateway to require you to call it back during the notifyUrl -- that usually happens during the returnUrl.  Normally the notifyUrl just needs to return a 200/OK response to indicate that the notification was successful.

I will take a look at your QuickPay driver and add any other comments there that I can think of.

**Comment by delatbabel** [comment]
[note 1]

This is not *always* subsequent to the redirect. The `purchase()` call can take some time and it's entirely possible for the POST to the notifyUrl to occur before the `purchase()` call returns. PaymentWall is an example of a gateway with this issue -- the `purchase()` call frequently takes as long as 45 seconds to complete, and the timeline of events frequently looks like this:

* Application calls `purchase()`
* PaymentWall send a POST to the notifyUrl with status = *approved*
* `purchase()` call returns with status = *error*

In this case the purchase must be assumed to be a success because the notifyUrl status takes precedence over the return response from `purchase()`.

Some further examples of the notifyUrl in use:

* MultiCards sends a very short response when a transaction fails.  It usually looks like this: status=3 (status 3 used for all failures) and message="error". That is all.  Subsequent to a failure they will send a much longer POST message to the notifyUrl containing all of the detailed reason for failure: fraud score, message from the customer's bank, more fine grained status code and error message, etc.
* AliPay (China) sends a sequence of up to 5 POST messages to the notifyUrl after a failed purchase giving detailed reasons for the failure.

**Comment by sanderha** [comment]
Thank you so much for all this information, it helped me understand omniapy *alot* better.

Also thanks for the PR you submitted with suggestions, I'll have a look ASAP
